# Nepal House Price Prediction Using Machine Learning

### Abstract

    Real estate pricing in Nepal remains largely informal and inconsistent,
    making it difficult for buyers, sellers, and investors to make informed
    decisions. This project presents an end-to-end machine learning solution
    for predicting house prices in Nepal using property listing data sourced
    from Hamrobazaar, a leading Nepali real estate platform.
    
    The dataset consists of over 2,200 property listings containing features
    such as land area (in Aana/Ropani), number of bedrooms and bathrooms,
    number of floors, road width and type, house facing direction, parking
    availability, city location, and available amenities. Extensive data
    preprocessing was performed to handle Nepal-specific land measurement
    units — converting Aana, Paisa, and Ropani into square feet — along with
    parsing of unstructured text fields, outlier removal, and log
    transformation of the skewed target variable (SalePrice in NPR).
    
    Exploratory data analysis revealed that house prices are heavily
    right-skewed with a skewness of 32.11, and that land area, number of
    floors, parking availability, and city location (Kathmandu commanding
    the highest premium) are the most influential predictors of price.
    
    Multiple regression models were evaluated including Ridge Regression as
    a baseline and XGBoost as the primary model. XGBoost achieved an R²
    score of 0.63 with an RMSE of approximately NPR 37.6 million on the
    test set, outperforming the Ridge baseline (R² = 0.46). A Scikit-learn
    Pipeline combining preprocessing and model inference was constructed for
    reproducibility and deployment readiness.
    
    The final model was integrated into a Django web application and deployed
    on Himalayan Hosting under the subdomain houseprice.aiwithprakas.com.np,
    enabling users to input property details and receive an instant price
    estimate. This project demonstrates the practical application of machine
    learning to address a real-world problem in the Nepali property market.
    
    Keywords: House Price Prediction, Nepal Real Estate, XGBoost, Regression,
              Machine Learning, Django, Hamrobazaar, Aana to Sqft Conversion

------------------

# नेपालमा घर मूल्य पूर्वानुमान (Machine Learning प्रयोग गरी)
### सारांश (Abstract)

नेपालमा घरजग्गाको मूल्य निर्धारण अझै पनि धेरै हदसम्म अनौपचारिक र असंगठित रहेको छ, जसका कारण खरिदकर्ता, विक्रेता तथा लगानीकर्ताहरूलाई सूचित निर्णय लिन गाह्रो हुन्छ। यस परियोजनाले नेपालको घर मूल्य पूर्वानुमान गर्नका लागि एक पूर्ण (end-to-end) मशीन लर्निङ समाधान प्रस्तुत गर्दछ, जसमा डेटा Hamrobazaar बाट संकलन गरिएको छ।

यस डेटासेटमा २,२०० भन्दा बढी घरजग्गाका विवरणहरू समावेश छन्, जसमा जमिनको क्षेत्रफल (आना/रोपनीमा), बेडरुम र बाथरुम संख्या, तल्लाको संख्या, बाटोको चौडाइ र प्रकार, घरको मुख फर्किएको दिशा, पार्किङ सुविधा, शहर स्थान, र उपलब्ध सुविधाहरू समावेश छन्। डेटा पूर्व-प्रशोधन (preprocessing) प्रक्रियामा नेपाल-विशेष जमिन मापन प्रणालीहरू (आना, पैसा, रोपनी) लाई स्क्वायर फिटमा रूपान्तरण गरिएको छ, साथै असंरचित टेक्स्ट डेटा सफा गरिएको, असामान्य मानहरू हटाइएको, र अत्यधिक झुकेको (skewed) लक्ष्य चर (SalePrice in NPR) लाई log transformation गरिएको छ।

डेटा विश्लेषणबाट देखियो कि घर मूल्य अत्यधिक दायाँतर्फ झुकेको (right-skewed) छ, जसको skewness 32.11 रहेको छ। साथै, जमिनको क्षेत्रफल, तल्लाको संख्या, पार्किङ सुविधा, र शहर स्थान (विशेष गरी काठमाडौँमा उच्च मूल्य) मूल्य निर्धारणका प्रमुख कारकहरू हुन्।

विभिन्न regression मोडेलहरू परीक्षण गरियो, जसमा Ridge Regression लाई आधार (baseline) को रूपमा र XGBoost लाई मुख्य मोडेलको रूपमा प्रयोग गरिएको थियो। XGBoost ले परीक्षण डेटामा R² स्कोर 0.63 र RMSE लगभग NPR 37.6 मिलियन प्राप्त गर्‍यो, जुन Ridge Regression (R² = 0.46) भन्दा राम्रो हो। साथै, Scikit-learn Pipeline प्रयोग गरी preprocessing र model inference लाई एकीकृत गरिएको छ, जसले पुन: प्रयोग (reproducibility) र deployment सजिलो बनाउँछ।

अन्तिम मोडेललाई Django web application मा समावेश गरी Himalayan Hosting मा houseprice.aiwithprakas.com.np subdomain मा deploy गरिएको छ, जसले प्रयोगकर्ताहरूलाई आफ्नो घरको विवरण इनपुट गरेर तुरुन्तै मूल्य अनुमान प्राप्त गर्न सक्षम बनाउँछ।

यो परियोजनाले नेपालको घरजग्गा बजारमा वास्तविक समस्याको समाधानका लागि मशीन लर्निङको व्यावहारिक प्रयोग देखाउँछ।

मुख्य शब्दहरू (Keywords):

घर मूल्य पूर्वानुमान, नेपाल रियल स्टेट, XGBoost, Regression, Machine Learning, Django, Hamrobazaar, आना देखि स्क्वायर फिट रूपान्तरण

----------------------

##  step 1  Import the necessary libraries and load the dataset.

In [4]:
import pandas as pd
import numpy as np

In [5]:
# read the dataset Nepal housing dataset
dataset = pd.read_csv("./Dataset/2020-4-27.csv")
dataset.head()

,Title,Address,City,Price,Bedroom,Bathroom,Floors,Parking,Face,Year,Views,Area,Road,Road Width,Road Type,Build Area,Posted,Amenities
0,Flat System House On Sale At Budhanikantha : H...,"Budhanikantha, Budhanilkantha, Kathmandu",Kathmandu,90000000,6,3,2.0,10,West,2073.0,17,1-0-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,18-0-0-0 Aana,11 hours ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
1,21 Aana Beautiful House : House For Sale In Bu...,"Pasikot, Budhanilkantha, Kathmandu",Kathmandu,80000000,5,3,2.0,9,East,2073.0,26,0-21-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-0-0 Aana,12 hours ago,"['Parking', 'Garage', 'Fencing', 'Frontyard', ..."
2,Beautiful To Floor House : House For Sale In D...,"Dhapasi, Dhapasi, Kathmandu",Kathmandu,70000000,5,3,2.0,12,East,2071.0,154,1-1-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,10-1-0-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
3,Modern Architecture Bungalow House : House For...,"baluwatar, Baluwatar, Kathmandu",Kathmandu,130000000,6,4,3.0,9,West,2017.0,17,0-10-1-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-7-1-0 Aana,1 day ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
4,Modern Bungalow House : House For Sale In Golf...,"Golfutar, Golfutar, Kathmandu",Kathmandu,100000000,6,3,2.0,10,East,2070.0,59,0-19-2-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-2-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Balcony', 'Back..."


* df.head  -check point
1. Price कुन column हो ?
2. कुन column text छ ?
3. कुन column number छ ?
4. Cleaning चाहिन्छ कि छैन ?
5. Missing value देखियो ?

In [4]:
# check the data info
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2211 entries, 0 to 2210
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Title       2211 non-null   object 
 1   Address     2211 non-null   object 
 2   City        2211 non-null   object 
 3   Price       2211 non-null   int64  
 4   Bedroom     2211 non-null   int64  
 5   Bathroom    2211 non-null   int64  
 6   Floors      1039 non-null   float64
 7   Parking     2211 non-null   int64  
 8   Face        2211 non-null   object 
 9   Year        582 non-null    float64
 10  Views       2211 non-null   object 
 11  Area        2211 non-null   object 
 12  Road        2211 non-null   object 
 13  Road Width  2211 non-null   object 
 14  Road Type   1426 non-null   object 
 15  Build Area  2211 non-null   object 
 16  Posted      2211 non-null   object 
 17  Amenities   2211 non-null   object 
dtypes: float64(2), int64(4), object(12)
memory usage: 311.0+ KB


` * The dataset consists of 18 columns, with Price as the dependent variable and the remaining 17 as independent variables. The dataset contains a total of 2,211 records.`

## Step 2:  Dataset ma House, Land, Apartment, Rent - Sabai mix xa . 
 Hami sirf "House For Sale" chahinchha, baaki hataayo. case=False le "house for sale", "HOUSE FOR SALE" duitai pakadchha.  

In [17]:
df = dataset[dataset['Title'].str.contains('House For Sale', case = False)]
df.head()


,Title,Address,City,Price,Bedroom,Bathroom,Floors,Parking,Face,Year,Views,Area,Road,Road Width,Road Type,Build Area,Posted,Amenities
0,Flat System House On Sale At Budhanikantha : H...,"Budhanikantha, Budhanilkantha, Kathmandu",Kathmandu,90000000,6,3,2.0,10,West,2073.0,17,1-0-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,18-0-0-0 Aana,11 hours ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
1,21 Aana Beautiful House : House For Sale In Bu...,"Pasikot, Budhanilkantha, Kathmandu",Kathmandu,80000000,5,3,2.0,9,East,2073.0,26,0-21-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-0-0 Aana,12 hours ago,"['Parking', 'Garage', 'Fencing', 'Frontyard', ..."
2,Beautiful To Floor House : House For Sale In D...,"Dhapasi, Dhapasi, Kathmandu",Kathmandu,70000000,5,3,2.0,12,East,2071.0,154,1-1-0-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,10-1-0-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
3,Modern Architecture Bungalow House : House For...,"baluwatar, Baluwatar, Kathmandu",Kathmandu,130000000,6,4,3.0,9,West,2017.0,17,0-10-1-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-7-1-0 Aana,1 day ago,"['Parking', 'Lawn', 'Garage', 'Frontyard', 'Fe..."
4,Modern Bungalow House : House For Sale In Golf...,"Golfutar, Golfutar, Kathmandu",Kathmandu,100000000,6,3,2.0,10,East,2070.0,59,0-19-2-0 Aana,20 Feet / Blacktopped,20 Feet,Blacktopped,0-14-2-0 Aana,2 days ago,"['Parking', 'Lawn', 'Garage', 'Balcony', 'Back..."


In [21]:
not_sale_df = dataset[~dataset['Title'].str.contains('House For Sale', case = False)]
not_sale_df.head()

,Title,Address,City,Price,Bedroom,Bathroom,Floors,Parking,Face,Year,Views,Area,Road,Road Width,Road Type,Build Area,Posted,Amenities
38,"Office Space For Rent In Radhe Radhe, Bhaktapur","Nearby Bhatbhateni Super market, Radhe Radhe, ...",Bhaktapur,50,0,2,3.5,2,West,2076.0,1.1K,2500 Sq. Feet,20 Feet / Blacktopped,20 Feet,Blacktopped,N/A Sq. Feet,4 months ago,"['Wifi', 'Drainage', 'Water Supply', 'Water Ta..."
95,"Plotted Land : Land For Sale In Sirutar, Bhakt...","Milan Chowk, Sirutar, Bhaktapur",Bhaktapur,2100000,0,0,NaN,0,North East,NaN,3.8K,0-5-2-0 Sq. Feet,20 Feet / Gravelled,20 Feet,Gravelled,N/A Sq. Feet,2 months ago,"['Drainage', 'Water Supply']"
101,Attractive Residential Land On Sale : Land For...,"Bhuddha colony, Budhanilkantha, Kathmandu",Kathmandu,3000000,0,0,NaN,0,West,NaN,48,12 Aana,16 Feet / Paved,16 Feet,Paved,N/A Aana,1 day ago,[]
102,18 Aana Land For Sale At Shamkhamul : Land For...,"Sankhamul, Sankhamul, Kathmandu",Kathmandu,6500000,0,0,NaN,0,East,NaN,46,0-18-0-0 Aana,40 Feet / Blacktopped,40 Feet,Blacktopped,N/A Aana,1 day ago,[]
103,Land On Sale At Beautiful Location : Land For ...,"High Vision Colony,Thankot,Mahadevsthan, Thank...",Kathmandu,2350000,0,0,NaN,0,South,NaN,232,5 Aana,13 Feet / Blacktopped,13 Feet,Blacktopped,N/A Aana,4 days ago,[]


In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1031 entries, 0 to 2210
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Title       1031 non-null   object 
 1   Address     1031 non-null   object 
 2   City        1031 non-null   object 
 3   Price       1031 non-null   int64  
 4   Bedroom     1031 non-null   int64  
 5   Bathroom    1031 non-null   int64  
 6   Floors      801 non-null    float64
 7   Parking     1031 non-null   int64  
 8   Face        1031 non-null   object 
 9   Year        484 non-null    float64
 10  Views       1031 non-null   object 
 11  Area        1031 non-null   object 
 12  Road        1031 non-null   object 
 13  Road Width  1031 non-null   object 
 14  Road Type   751 non-null    object 
 15  Build Area  1031 non-null   object 
 16  Posted      1031 non-null   object 
 17  Amenities   1031 non-null   object 
dtypes: float64(2), int64(4), object(12)
memory usage: 153.0+ KB


-----------
##  step3  Convert to number :
 'Parse Functions  yo sabai bata important step ho - raw messay data lai numbers ma convert gariyo.
 
 parse_view - "2.6k -> 2600
 


In [41]:
df.Views.unique()[30:100]

array(['157', '566', '96', '345', '971', '151', '282', '66', '339', '71',
       '136', '108', '150', '112', '1.7K', '100', '67', '161', '139',
       '168', '5.6K', '274', '1.4K', '209', '162', '588', '437', '726',
       '688', '374', '173', '599', '205', '677', '5.1K', '6.2K', '188',
       '86', '134', '710', '166', '602', '792', '94', '197', '340', '220',
       '380', '135', '325', '960', '673', '167', '296', '2.1K', '917',
       '8.2K', '4.1K', '1.9K', '138', '5.3K', '206', '103', '214', '453',
       '247', '786', '461', '434', '290'], dtype=object)

In [45]:
df['Views_num'] = (
    df['Views']
    .astype(str)
    .str.upper()
    .str.replace('K', 'e3')
    .str.replace('M', 'e6')
)

df['Views_num'] = pd.to_numeric(
    df['Views_num'],
    errors='coerce'
)

C:\Users\s230250\AppData\Local\Temp\ipykernel_58672\869659580.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Views_num'] = (
C:\Users\s230250\AppData\Local\Temp\ipykernel_58672\869659580.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Views_num'] = pd.to_numeric(


In [56]:
df['Area_sqft'].info

<bound method Series.info of 0        5476.000
1        7187.250
2        5818.250
3        3508.060
4        6673.870
          ...    
2203    18139.250
2204    10267.500
2207      855.625
2208     2738.000
2210     4449.250
Name: Area_sqft, Length: 1031, dtype: float64>

## step 4 Convert Ann to Sqft 
     parse_area - "1-0-0-0 Anna" -> sqft
     Nepal ma land Aana/Ropani ma mapinchha — model lai sqft chahinchha. Conversion rules:

     1 Ropani = 16 Aana  = 5476 sqft
    1 Aana   = 4 Paisa  = 342.25 sqft
    1 Paisa  = 4 Dam    = 85.56 sqft
    1 Dam    =            21.39 sqft

    Format "1-0-0-0 Aana" ko matlab:
        1 Ropani, 0 Aana, 0 Paisa, 0 Dam
            = 1×16×342.25 = 5476 sqft

        re.sub(r'(\d)\.(\d)', r'\1-\2', s) le "3.1.0 Aana" jasto dot-separated format lai "3-1-0 Aana" ma convert garcha — kinabhane dataset ma inconsistent format thiyo.

In [53]:
df.Area

0         1-0-0-0 Aana
1        0-21-0-0 Aana
2         1-1-0-0 Aana
3        0-10-1-0 Aana
4        0-19-2-0 Aana
             ...      
2203    3-5-0-0 Ropani
2204           30 Aana
2207          2.5 Aana
2208            8 Aana
2210           13 Aana
Name: Area, Length: 1031, dtype: object

In [50]:
import re
def parse_area(s):
    #-----------------------------------------
    # parse_area - "1-0-0-0 Anna" -> sqft
    # Nepal ma land anna/Ropani ma napinxa  
    # hamile teslai sqrt ma change garnu parxa 
    # -----------------------------------------------
    s = str(s).strip()
    # Normalize dot-separated numbers (e.g. "3.1.0 Aana" → "3-1-0 Aana")
    s2 = re.sub(r'(\d)\.(\d)', r'\1-\2', s)
 
    # R-A-P-D Aana (4 parts)
    m = re.match(r'^(\d+)-(\d+)-(\d+)-(\d+)\s*/?\s*[\d.]*\s*Aana', s2, re.I)
    if m:
        r, a, p, d = int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))
        return r * 16 * 342.25 + a * 342.25 + p * 85.56 + d * 21.39
    # R-A-P Aana (3 parts)
    m = re.match(r'^(\d+)-(\d+)-(\d+)\s*/?\s*[\d.]*\s*Aana', s2, re.I)
    if m:
        r, a, p = int(m.group(1)), int(m.group(2)), int(m.group(3))
        return r * 16 * 342.25 + a * 342.25 + p * 85.56
    # X Aana / X ana Aana
    m = re.match(r'^([\d.]+)\s*(?:ana\s*|aana\s*)?Aana', s, re.I)
    if m:
        try: return float(m.group(1)) * 342.25
        except: return np.nan
    # R-A-P-D Ropani
    m = re.match(r'^(\d+)-(\d+)-(\d+)-(\d+)\s*Ropani', s2, re.I)
    if m:
        r, a, p, d = int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))
        return r * 5476 + a * 342.25 + p * 85.56 + d * 21.39
    # X Ropani
    m = re.match(r'^([\d.]+)\s*Ropani', s, re.I)
    if m: return float(m.group(1)) * 5476
    # X Sq. Feet
    m = re.match(r'^([\d.]+)\s*Sq', s, re.I)
    if m: return float(m.group(1))
    return np.nan
 
df['Area_sqft'] = df['Area'].apply(parse_area)

C:\Users\s230250\AppData\Local\Temp\ipykernel_58672\1982536482.py:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Area_sqft'] = df['Area'].apply(parse_area)


In [52]:
df['Area_sqft'].unique()[:30]

array([ 5476.   ,  7187.25 ,  5818.25 ,  3508.06 ,  6673.87 ,  4385.07 ,
       16855.81 ,  3165.8  ,  4406.46 ,  1369.   ,  4791.5  ,  1540.12 ,
        1711.25 ,  2738.   ,  2053.5  ,  3422.5  ,  1197.87 ,  6502.75 ,
        2087.725,  2224.625,  1026.75 , 16599.12 , 21904.   ,  2395.75 ,
        1060.975,  3764.75 , 10267.5  ,  3080.25 ,  1197.875,  4107.   ])

In [57]:
df['Area_sqft'].info

<bound method Series.info of 0        5476.000
1        7187.250
2        5818.250
3        3508.060
4        6673.870
          ...    
2203    18139.250
2204    10267.500
2207      855.625
2208     2738.000
2210     4449.250
Name: Area_sqft, Length: 1031, dtype: float64>

---------------------------
## Step 5 Road Road Width
    20 feet / backtopped bata 20 matra nikalera 
    Meter aayo bhane x 3.281 gari feet ma convert garne 



In [61]:
df['Road Width']

0       20 Feet 
1       20 Feet 
2       20 Feet 
3       20 Feet 
4       20 Feet 
          ...   
2203     12 Feet
2204     30 Feet
2207     13 Feet
2208     12 Feet
2210     12 Feet
Name: Road Width, Length: 1031, dtype: object

In [58]:
# --- Road Width: "20 Feet / Blacktopped" → 20 ---
def parse_road_width(rw):
    m = re.search(r'([\d.]+)\s*(Feet|Meter|feet|meter)', str(rw))
    if m:
        val = float(m.group(1))
        if 'meter' in m.group(2).lower():
            val *= 3.281  # meter to feet
        return val
    return np.nan

In [62]:
df['RoadWidth_ft'] = df['Road Width'].apply(parse_road_width)

C:\Users\s230250\AppData\Local\Temp\ipykernel_58672\27587905.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['RoadWidth_ft'] = df['Road Width'].apply(parse_road_width)


In [63]:
df['RoadWidth_ft'] 

0       20.0
1       20.0
2       20.0
3       20.0
4       20.0
        ... 
2203    12.0
2204    30.0
2207    13.0
2208    12.0
2210    12.0
Name: RoadWidth_ft, Length: 1031, dtype: float64

----------------------------------------
## Step 6 Roade type  String to Rank number 
    road_rank = {'Blacktopped':4, 'Concrete':3, 'Paved':2, 'Gravelled':1}
    *  Blacktopped = sabai ramro road = 4
    * Gravelled   = kachchi road     = 1

In [65]:
df['Road Type']

0        Blacktopped
1        Blacktopped
2        Blacktopped
3        Blacktopped
4        Blacktopped
            ...     
2203             NaN
2204             NaN
2207             NaN
2208             NaN
2210             NaN
Name: Road Type, Length: 1031, dtype: object

In [66]:
# --- Road Type: string → rank ---

road_rank = {'Blacktopped': 4, 'Concrete': 3, 'Paved': 2, 'Gravelled': 1}
def parse_road_type(rt):
    for k, v in road_rank.items():
        if k.lower() in str(rt).lower():
            return v
    return 0

In [71]:
dataset.isnull().sum()

Title            0
Address          0
City             0
Price            0
Bedroom          0
Bathroom         0
Floors        1172
Parking          0
Face             0
Year          1629
Views            0
Area             0
Road             0
Road Width       0
Road Type      785
Build Area       0
Posted           0
Amenities        0
dtype: int64

In [46]:
df['Views_num']

0          17.0
1          26.0
2         154.0
3          17.0
4          59.0
         ...   
2203    15000.0
2204     5600.0
2207     1400.0
2208     1500.0
2210    19600.0
Name: Views_num, Length: 1031, dtype: float64

In [11]:
# percentage of missing value 
dataset.isnull().sum()/dataset.shape[0]*100 

Title          0.000000
Address        0.000000
City           0.000000
Price          0.000000
Bedroom        0.000000
Bathroom       0.000000
Floors        53.007689
Parking        0.000000
Face           0.000000
Year          73.677069
Views          0.000000
Area           0.000000
Road           0.000000
Road Width     0.000000
Road Type     35.504297
Build Area     0.000000
Posted         0.000000
Amenities      0.000000
dtype: float64

In [13]:
for col in dataset.columns:
    if dataset[col].isnull().sum()>0:
        print(col)

Floors
Year
Road Type


In [17]:
# step 1
# Missing ra Non-missing grop bich ma Price ma difference xa ki xaina ?
missing_floors = dataset[dataset['Floors'].isnull()]['Price']
non_missing_floors = dataset[dataset['Floors'].notnull()]['Price']
print("missing_floors_mean:",missing_floors.mean())
print("non_missing_floors_mean: ", non_missing_floors.mean())

In [31]:
# step 2 
from scipy.stats import mannwhitneyu
stas,p = mannwhitneyu(missing_floors, non_missing_floors)
if p < 0.05:
    print( "Significant Difference -> missing important - add indicator, missing हुनु र Price बीच सम्बन्ध छ (missing important छ) ")
else :
    print("Not significant - missing random-Drop/fill")
    

Significant Difference -> missing important - add indicator, missing हुनु र Price बीच सम्बन्ध छ (missing important छ) 


##### floor Missing huda price High xa 
・ yo difference real ho ki chance le matra ayeko ho test garam

In [3]:
# Example DataFrame
from scipy.stats import ttest_ind
df = dataset.copy()
columns_to_check = ['Floors', 'Year', 'Road Type']  # Missing check गर्ने columns
target_col = 'Price'                        # Target column

# Result store गर्ने list
results = []

for col in columns_to_check:
    missing_group = df[df[col].isnull()][target_col]
    non_missing_group = df[df[col].notnull()][target_col]
    
    # Check if both groups have enough data
    if len(missing_group) > 1 and len(non_missing_group) > 1:
        t_stat, p_value = mannwhitneyu(missing_group, non_missing_group)
        missing_avg = missing_group.mean()
        non_missing_avg = non_missing_group.mean()
        diff = abs(missing_avg - non_missing_avg)
        
        results.append({
            'Column': col,
            'Missing Avg': missing_avg,
            'Non-Missing Avg': non_missing_avg,
            'Diff': diff,
            'T-stat': t_stat,
            'P-value': p_value
        })
    else:
        results.append({
            'Column': col,
            'Missing Avg': np.nan,
            'Non-Missing Avg': np.nan,
            'Diff': np.nan,
            'T-stat': np.nan,
            'P-value': np.nan
        })

# Results देखाउने
results_df = pd.DataFrame(results)
print(results_df)
    

NameError: name 'mannwhitneyu' is not defined

In [40]:
for col, p_value in zip(results_df['Column'], results_df['P-value']):
    if p_value < 0.05:
        print(f"{col} -> Significant Difference: missing important, add indicator")
    else:
        print(f"{col} -> Not Significant: drop or impute")

Floors -> Significant Difference: missing important, add indicator
Year -> Significant Difference: missing important, add indicator
Road Type -> Significant Difference: missing important, add indicator


In [19]:
not_sale_df= dataset[~dataset['Title'].str.contains('House For Sale', case = False)]
not_sale_df

,Title,Address,City,Price,Bedroom,Bathroom,Floors,Parking,Face,Year,Views,Area,Road,Road Width,Road Type,Build Area,Posted,Amenities
38,"Office Space For Rent In Radhe Radhe, Bhaktapur","Nearby Bhatbhateni Super market, Radhe Radhe, ...",Bhaktapur,50,0,2,3.5,2,West,2076.0,1.1K,2500 Sq. Feet,20 Feet / Blacktopped,20 Feet,Blacktopped,N/A Sq. Feet,4 months ago,"['Wifi', 'Drainage', 'Water Supply', 'Water Ta..."
95,"Plotted Land : Land For Sale In Sirutar, Bhakt...","Milan Chowk, Sirutar, Bhaktapur",Bhaktapur,2100000,0,0,NaN,0,North East,NaN,3.8K,0-5-2-0 Sq. Feet,20 Feet / Gravelled,20 Feet,Gravelled,N/A Sq. Feet,2 months ago,"['Drainage', 'Water Supply']"
101,Attractive Residential Land On Sale : Land For...,"Bhuddha colony, Budhanilkantha, Kathmandu",Kathmandu,3000000,0,0,NaN,0,West,NaN,48,12 Aana,16 Feet / Paved,16 Feet,Paved,N/A Aana,1 day ago,[]
102,18 Aana Land For Sale At Shamkhamul : Land For...,"Sankhamul, Sankhamul, Kathmandu",Kathmandu,6500000,0,0,NaN,0,East,NaN,46,0-18-0-0 Aana,40 Feet / Blacktopped,40 Feet,Blacktopped,N/A Aana,1 day ago,[]
103,Land On Sale At Beautiful Location : Land For ...,"High Vision Colony,Thankot,Mahadevsthan, Thank...",Kathmandu,2350000,0,0,NaN,0,South,NaN,232,5 Aana,13 Feet / Blacktopped,13 Feet,Blacktopped,N/A Aana,4 days ago,[]
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2198,"Sun City : Apartment For Sale In Pepsicola, Ka...","Pepsicola, Pepsicola, Kathmandu",Kathmandu,11000000,3,2,6.0,1,North West,2071.0,16.1K,1336 Sq. Feet,20 Feet,20 Feet,NaN,N/A Sq. Feet,1 year ago,"['Backyard', 'Balcony', 'Frontyard', 'Garden',..."
2199,"Land For Sale In Gothatar, Kathmandu","gothatar kadaghari, Gothatar, Kathmandu",Kathmandu,2500000,0,0,NaN,0,North,NaN,1.3K,0-4-2-2 Aana,8 Meter,8 Meter,NaN,N/A Sq. Feet,1 year ago,[]
2205,"Land For Sale In Budhanilkantha, Kathmandu","Deuba Chowk, Budhanilkantha, Kathmandu",Kathmandu,2700000,0,0,NaN,0,North,NaN,7.8K,7-0-2 Aana,12 Feet,12 Feet,NaN,N/A Sq. Feet,1 year ago,"['Drainage', 'Water Supply']"
2206,"House For Rent In Chapali, Kathmandu","Bhangal, Chapali, Kathmandu",Kathmandu,500000,5,5,3.0,5,North East,NaN,3.1K,1 Ropani,16 Feet,16 Feet,NaN,N/A Sq. Feet,1 year ago,"['Balcony', 'Frontyard', 'Garden', 'Drainage',..."
